lib/job-iteration/nested_enumerator.rb

```diff
@@ -0,0 +1,40 @@
+# frozen_string_literal: true
+
+module JobIteration
+  # @private
+  class NestedEnumerator
+    def initialize(enums, cursor: nil)
+      unless enums.all?(Proc)
+        raise ArgumentError, "enums must contain only procs/lambdas"
+      end
+
+      if cursor && enums.size != cursor.size
+        raise ArgumentError, "cursor should have one item per enum"
+      end
+
+      @enums = enums
+      @cursor = cursor || Array.new(enums.size)
+    end
+
+    def each(&block)
+      return to_enum unless block_given?
+
+      iterate([], [], 0, &block)
+    end
+
+    private
+
+    def iterate(current_items, current_cursor, index, &block)
+      cursor = @cursor[index]
+      enum = @enums[index].call(*current_items, cursor)
+
+      enum.each do |item, cursor_value|
+        if index == @cursor.size - 1
+          yield item, current_cursor + [cursor_value]
```

Comment:
I've been stepping through the debugger a few times and what stood out to me is `cursor = @cursor[index]`, the initial cursor value stays 'stuck' for subsequent iterations. In my array example this means skipping over `"2a"` but also `"3a"` if you extend the pattern as such:

```ruby
integers = [0, 1, 2, 3]
strings = [["0a", "0b", "0c"], ["1a", "1b", "1c"], ["2a", "2b", "2c"], ["3a", "3b", "3c"]]
```

This seems to do the trick unless I'm missing something (it's getting late here)?
```suggestion
          yield item, current_cursor + [cursor_value]
          @cursor[index] = nil
```

Reply:
I guess what I am not grasping is why the cursor logic throughout this library needs to change to make nesting work. AFAIK its `nil` to start at the beginning of a collection, and a non-nil cursor to resume _after_. With the generic way nested enumerators are set up I don't see why this should change? The individual enumerators shouldn't need to care if it's nested or not, the nested enumerator should feed the correct value to them and handle the nesting level (`index`).

To whiteboard the possibly cursor values for `["0a", "0b", "0c"]` and the item processed in each_iteration:
- cursor: `nil`, item: "0a"
- cursor: `0`, item: "0b"
- cursor: `1`, item: "0c"

and nested with: `[["0a", "0b", "0c"], ["1a", "1b", "1c"], ["2a", "2b", "2c"]]`:
- cursor: `[nil, nil]`, item: "0a"
- cursor: `[nil, 0]`, item: "0b"
- cursor: `[nil, 1]`, item: "0c"
- cursor: `[0, nil]`, item: "1a"
- cursor: `[0, 0]`, item: "1b"
- cursor: `[0, 1]`, item: "1c"
- cursor: `[1, nil]`, item: "2a"

etc. It seems to me resetting the right cursor index to nil at the correct time should make the above possible and make it work without changes to the rest of the library.

Having debugged some more I think it's also possible to refactor out `current_cursor` in the iterate method, it seems redundant with `@cursor` being initialized (from the interrupted cursor or an array of nils) but never updated 🤔 I'll try to get a PR up in the next few days. 

Either way - thanks for taking the time to work on this and answering my questions 🙇‍♂️